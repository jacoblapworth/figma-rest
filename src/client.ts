import got from 'got'
import type { Got } from 'got'

import { API_PROTOCOL, API_HOST, API_VER } from './config.js'
import { VariablesResource } from './resources/variables.js'
import { ProjectsResource } from './resources/projects.js'
import { CommentsResource } from './resources/comments.js'
import { FilesResource } from './resources/files.js'
import { UsersResource } from './resources/users.js'
import { VersionsResource } from './resources/versions.js'
import { ActivityLogsResource } from './resources/activityLogs.js'
import { PaymentsResource } from './resources/payments.js'
import { DevResourcesResource } from './resources/devResources.js'

type PersonalConfig = {
  personalAccessToken: string
}

type OAuthConfig = {
  oAuthToken: string
}

type SharedConfig = {
  apiVersion?: string
  protocol?: string
  host?: string
}

type ClientConfig = SharedConfig & (PersonalConfig | OAuthConfig)

export class FigmaClient {
  request: Got

  personalAccessToken?: string
  oAuthToken?: string

  activityLogs: ActivityLogsResource
  comments: CommentsResource
  devResources: DevResourcesResource
  files: FilesResource
  payments: PaymentsResource
  projects: ProjectsResource
  users: UsersResource
  variables: VariablesResource
  versions: VersionsResource

  constructor({
    apiVersion = API_VER,
    protocol = API_PROTOCOL,
    host = API_HOST,
    ...config
  }: ClientConfig) {
    this.request = got.extend({
      prefixUrl: new URL(`${protocol}://${host}`),
      responseType: 'json',
      resolveBodyOnly: true,
    })

    if ('personalAccessToken' in config) {
      this.personalAccessToken = config.personalAccessToken
      this.request = this.request.extend({
        headers: { 'X-Figma-Token': this.personalAccessToken },
      })
    }

    if ('oAuthToken' in config) {
      this.oAuthToken = config.oAuthToken
      this.request = this.request.extend({
        headers: { Authorization: `Bearer ${this.oAuthToken}` },
      })
    }

    this.activityLogs = new ActivityLogsResource(this)
    this.comments = new CommentsResource(this)
    this.devResources = new DevResourcesResource(this)
    this.files = new FilesResource(this)
    this.payments = new PaymentsResource(this)
    this.projects = new ProjectsResource(this)
    this.users = new UsersResource(this)
    this.variables = new VariablesResource(this)
    this.versions = new VersionsResource(this)
  }
}
