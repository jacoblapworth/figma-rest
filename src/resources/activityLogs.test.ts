import { describe, expect, test } from 'vitest'
import { FigmaClient } from '../client.js'

describe('variables', () => {
  const client = new FigmaClient({ personalAccessToken: 'abc' })

  test('logs', async () => {
    const result = await client.activityLogs.get()
    expect(result).toMatchSnapshot()
  })
})
