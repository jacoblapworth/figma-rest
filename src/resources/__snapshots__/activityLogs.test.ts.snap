// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`variables > logs 1`] = `
{
  "done": false,
  "value": {
    "action": {
      "details": {
        "figjam_paid_status": "viewer",
        "figma_paid_status": "viewer",
        "permission": "member",
      },
      "type": "org_user_create",
    },
    "actor": {
      "email": "admin@figma.com",
      "id": "1099091282752443416",
      "name": "Admin",
      "type": "user",
    },
    "context": {
      "ip_address": "172.19.0.1",
      "is_figma_support_team_action": false,
      "org_id": "1047918802483077121",
    },
    "entity": {
      "email": "member@figma.com",
      "id": "1099091282712783786",
      "name": "Member",
      "type": "user",
    },
    "id": "1243",
    "timestamp": 1650578182,
  },
}
`;
